sys/kern/dev_procstat.c

```diff
@@ -0,0 +1,208 @@
+#include <sys/malloc.h>
+#include <stdio.h>
+
+/* Implementation of /dev/procstat
+ *
+ * When /dev/procstat is opened it takes a snapshot of info of all existing
+ * processes.
+ *
+ * During read calls to /dev/procstat this info can be read.
+ *
+ * Example:
+ * euid   pid    ppid    pgrp   session  state    elfpath
+ * 0       1       0       1       1       R      /bin/ksh
+ */
+
+/* maximum size of output string of proc info
+ * name + path + state (1 character)+ spaces (10 characters) +
+ *  + 5 * max length of uint32 (10 characters)*/
+#define MAX_P_STRING (PATH_MAX + 5 * 10 + 1 + 10)
+
+/* maximum amount of processes that procstat can handle */
+#define MAX_PROC 40
+
+/* we want to have at most 10 instances of procstat */
+#define MAX_PROCSTAT 10
+
+static char proc_state[4] = {
+  [PS_NORMAL] = 'R',
+  [PS_STOPPED] = 'S',
+  [PS_DYING] = 'D',
+  [PS_ZOMBIE] = 'Z',
+};
+
+typedef struct ps_entry {
+  pid_t pid;
+  pid_t ppid;
+  uid_t uid;
+  pgid_t pgrp;
+  sid_t sid;
+  proc_state_t proc_state;
+  char *elfpath;
+} ps_entry_t;
+
+typedef struct ps_buf {
+  mtx_t lock;
+  int nproc;
+  int cur_proc;
+  size_t offset; /* offset in buffer of current proc string */
+  size_t psize;  /* size of current proc string */
+  char buf[MAX_P_STRING];
+  ps_entry_t ps[];
+} ps_buf_t;
+
+static int ps_opencnt = 0;
+static mtx_t procstat_lock;
+
+static int dev_procstat_read(file_t *f, uio_t *uio);
+static int dev_procstat_close(vnode_t *v, file_t *fp);
+static int dev_procstat_open(vnode_t *v, int mode, file_t *fp);
+
+static fileops_t dev_procstat_fileops = {
+  .fo_read = dev_procstat_read,
+  .fo_close = default_vnclose,
+  .fo_stat = default_vnstat,
+  .fo_ioctl = default_vnioctl,
+};
+
+static vnodeops_t dev_procstat_vnodeops = {
+  .v_open = dev_procstat_open,
+  .v_close = dev_procstat_close,
+};
+
+static void ps_entry_fill(ps_entry_t *pe, proc_t *p) {
+  SCOPED_MTX_LOCK(&p->p_lock);
+  pe->uid = p->p_cred.cr_euid;
+  pe->pid = p->p_pid;
+  pe->ppid = p->p_parent->p_pid;
+  pe->pgrp = p->p_pgrp->pg_id;
+  pe->sid = p->p_pgrp->pg_session->s_sid;
+  pe->proc_state = p->p_state;
+  pe->elfpath = kstrndup(M_TEMP, p->p_elfpath, PATH_MAX);
+}
+
+/* buf must be at least MAX_P_STRING long
+ * returns length of written string
+ */
+static int ps_entry_tostring(char *buf, ps_entry_t *pe) {
+  int r = snprintf(buf, MAX_P_STRING, "%d\t%d\t%d\t%d\t%d\t%c\t%s\n", pe->uid,
+                   pe->pid, pe->ppid, pe->pgrp, pe->sid,
+                   proc_state[pe->proc_state], pe->elfpath);
+
+  return MIN(r, MAX_P_STRING);
+}
+
+static int dev_procstat_open(vnode_t *v, int mode, file_t *fp) {
+  WITH_MTX_LOCK (&procstat_lock) {
+    if (ps_opencnt >= MAX_PROCSTAT)
+      return EMFILE;
+    ps_opencnt++;
+    assert(ps_opencnt <= MAX_PROCSTAT);
+  }
+
+  int error = 0;
+  proc_t *p;
+  ps_buf_t *ps;
+
+  if ((error = vnode_open_generic(v, mode, fp)))
+    goto err;
+
+  ps = kmalloc(M_TEMP, MAX_PROC * sizeof(ps_entry_t) + sizeof(ps_buf_t), 0);
```

Comment:
Actually, I looked at `kmalloc()` and it looks like it can't return `NULL` if you don't pass `M_NOWAIT`, so there's no need to check for `NULL` here. This should simplify things a bit.

Reply:
I don't think it is a good idea to drop this check. I think that in future it could be such a case and it would be hard to find this line if something braked.

Reply:
I think most if not all uses of `kmalloc()` without `M_NOWAIT` assume that the return value is not `NULL`, so I think the chances of that changing are near-zero.

Reply:
Okay, I changed it.